edition = "2024";

import "tokens.proto";

package ast_protobuf;

message SerializedAstNode
{
  repeated SerializedAstNode children = 1;

  oneof token 
  {
    WhileToken while_token = 10;
    IfToken if_token = 11;
    PrintToken print_token = 12;
    InputToken input_token = 13;
    IdentifierToken identifier_token = 14;
    LiteralToken literal_token = 15;
    AssignmentToken assignment_token = 16;
    AddAssignmentToken add_assignment_token = 17;
    SubAssignmentToken sub_assignment_token = 18;
    MulAssignmentToken mul_assignment_token = 19;
    DivAssignmentToken div_assignment_token = 20;
    ModAssignmentToken mod_assignment_token = 21;
    AddToken add_token = 22;
    SubToken sub_token = 23;
    MulToken mul_token = 24;
    DivToken div_token = 25;
    ModToken mod_token = 26;
    EqualToken equal_token = 27;
    NotEqualToken not_equal_token = 28;
    LessToken less_token = 29;
    LessEqualToken less_equal_token = 30;
    GreaterToken greater_token = 31;
    GreaterEqualToken greater_equal_token = 32;
    UnaryPlusToken unary_plus_token = 33;
    UnaryMinusToken unary_minus_token = 34;
    PrefixIncrementToken prefix_increment_token = 35;
    PostfixIncrementToken postfix_increment_token = 36;
    PrefixDecrementToken prefix_decrement_token = 37;
    PostfixDecrementToken postfix_decrement_token = 38;
  }
}

message SerializedAstRoot
{
  repeated SerializedAstNode children = 1;
}

